File: 动态规划/746.使用最小花费爬楼梯.go
//记忆化搜索
func minCostClimbingStairs(cost []int) int {
	n := len(cost)
	memo := make([]int, n+1)
	for i := 0; i <= n; i++ {
		memo[i] = -1
	}
	return dfs(cost, n, memo)
}

func dfs(cost []int, i int, memo []int) int {
	if i < 2 {
		return 0
	}
	if memo[i] != -1 {
		return memo[i]
	}
	memo[i] = min(dfs(cost, i-1, memo)+cost[i-1], dfs(cost, i-2, memo)+cost[i-2])
	return memo[i]
}

//递推
func minCostClimbingStairs1(cost []int) int {
	n := len(cost)
	fn := make([]int, n+1)
	for i := 2; i <= n; i++ {
		fn[i] = min(fn[i-1]+cost[i-1], fn[i-2]+cost[i-2])
	}
	return fn[n]
}

File: 动态规划/746.使用最小花费爬楼梯.java
class Solution {
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] memo = new int[n+1];
        Arrays.fill(memo,-1);
        return dfs(cost,memo,n);
    }
    public int dfs(int[] cost,int[] memo,int i){
        if(i<2){
            return 0;
        }
        if(memo[i]!=-1){
            return memo[i];
        }
        return memo[i]=Math.min(cost[i-1]+dfs(cost,memo,i-1),cost[i-2]+dfs(cost,memo,i-2));
    }
}